### DISCUSSION TITLE
Queries executed using `em.getKnex().raw` not getting rolled back

### DISCUSSION THREAD

**Discussion by stad-nico** [mikro-orm/mikro-orm#5294] [General]
### Describe the bug

Queries executed on an `EntityManager` instance using `getKnex().raw()` are always auto-committed (or implicitly committed?) and thus rolling back wont work either way: starting a transaction using `em.begin()` or using `em.transactional(() => ...)`.

### Reproduction

```ts
await entityManager.transactional(async (em) => {
    await em.getKnex('write').raw("INSERT INTO directories (name) VALUES ('test')");
    throw new Error('rollback');
});
```
or
```ts
await entityManager.begin();
await em.getKnex('write').raw("INSERT INTO directories (name) VALUES ('test')");
await entityManager.rollback();
```

### What driver are you using?

@mikro-orm/mariadb

### MikroORM version

6.1.4

### Node.js version

19.6.0 typescript 5.3.3

### Operating system

Windows 10

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
You should be doing this: 

```ts
await entityManager.transactional(async (em) => {
    await em.execute("INSERT INTO directories (name) VALUES ('test')");
    throw new Error('rollback');
});
```

Only that way you are using the same connection. Using `em.getKnex()` will not respect your current transaction context, that's just a getter for knex instance available on the connection level (its a shortcut for `this.getConnection(type).getKnex()`).

Alternatively, provide the context to knex query:

```ts
await entityManager.transactional(async (em) => {
    await em.getKnex('write')
        .raw("INSERT INTO directories (name) VALUES ('test')")
        .transacting(em.getTransactionContext()!) // <-- here
    throw new Error('rollback');
});
```

Note that you can't set a transaction context to the knex instance itself, its an option of knex query builder, not the knex instance.

**Reply by stad-nico** [reply]
Using `em.execute` is not an option for me, I am executing raw sql statements and need the `getKnex().raw<T>()` generic argument to specify the type of the returned entity from the db call.

However using `.transacting` works perfectly with raw sql and knex\`s `raw` method.

But I have to admit it was quite an unpleasant DX getting all that to work for unit tests with nested transactions using the explicit `begin` and `commit` functions. I eventually got it to work 

```ts
beforeAll(async () => {
    await globalEntityManager.begin();
    globalTransactionContext = globalEntityManager.getTransactionContext();
});

afterAll(async () => {
    await globalEntityManager.getConnection().rollback(globalTransactionContext);
});

beforeEach(async () => {
    await globalEntityManager.begin({ ctx: globalTransactionContext });
    currentTransactionContext = globalEntityManager.getTransactionContext();
});

afterEach(async () => {
    await globalEntityManager.getConnection().rollback(currentTransactionContext);
});

describe('test', () => {
    it('should rollback everything', async () => {
        await globalEntityManager
	    .getKnex()
	    .raw("INSERT INTO directories (name) VALUES ('test')")
	    .transacting(currentTransactionContext!);
    });
});
```

I am not sure whether this is the correct way as I could not find anything in the docs related to nested transactions not using `em.transactional`, so it was trail and error when to use `.getConnection()` to achieve the correct behaviour. 
Maybe some docs on that topic would be beneficial, however I dont know how often people need nested transaction (and why they wouldnt use `em.transactional` expect for specific use cases) so maybe this is rather unimportant.

It also seems over complicated having to get the current transaction context after calling `em.begin({ctx: ...})` to get the current context for the inner transaction and even more odd having to call `em.getConnection().rollback(currentTransactionContext)` to rollback that inner transaction.

Unfortunately I cannot tell if this would suit the way mikro orm is internally design but for me it would feel the most natural if `em.begin({ctx: ...})` simply returned a new `EntityManager` scoped to the current transaction context. Calling `commit` and `rollback` on this instance would then simply commit or rollback the inner transaction. 
Maybe this would make switching between `em.transactional` and `em.begin` and `em.rollback/em.commit` easier and IMO would fit better with how `em.transactional` is designed as it also provides a new `EntityManager` instance for the callback function.

**Reply by B4nan** [reply]
> Using em.execute is not an option for me, I am executing raw sql statements and need the getKnex().raw<T>() generic argument to specify the type of the returned entity from the db call.

You can pass a knex qb instance to `em.execute()` too. If you want logging to work, you should use `em.execute()` even for knex instances.

> I am not sure whether this is the correct way as I could not find anything in the docs related to nested transactions not using em.transactional, so it was trail and error when to use .getConnection() to achieve the correct behaviour.

You shouldn't be using the connection methods at all, that's rather internal, the EM has `begin/commit/rollback` methods, and those respect the transaction context (begin stores it, commit/rollback reuses it). And instead of trying to commit/rollback transactions from the global EM instance via the connection, you should create a fork and use the methods on that - which will again respect the transaction context you create. The fork will represent the transaction context, so you no longer need that variable. Its kinda the same as what you did, but with using the API as they were designed instead of using the internals:

```ts
let globalEntityManager: EntityManager;
let fork: EntityManager;

beforeAll(async () => {
  await globalEntityManager.begin();
});

afterAll(async () => {
  await globalEntityManager.commit();
});

beforeEach(async () => {
  fork = globalEntityManager.fork();
  await fork.begin();
});

afterEach(async () => {
  await fork.rollback();
});

describe('test', () => {
  it('should rollback everything', async () => {
    const query = fork.getKnex().raw("INSERT INTO directories (name) VALUES ('test')"); // this can be the global one as well, its not important
    await fork.execute(query);
  });
});
```

**Reply by stad-nico** [reply]
Thank you, this works like a charm! 

This API is way more intuitiv than the workaround I threw together and is also documented under [Transactions and Concurrency > Approach 2: Explicitly](https://mikro-orm.io/docs/transactions#approach-2-explicitly), I seem to have missed that.


> You can pass a knex qb instance to `em.execute()` too. If you want logging to work, you should use `em.execute()` even for knex instances.

This however doesn't seem to preserve the return type. Annotating the return type in a knex qb still gives me `EntityData<Partial<any>>[]` when using `em.execute`.



> You shouldn't be using the connection methods at all, that's rather internal

Is there any way to close the connection after all unit tests without using `em.getConnection().close()` if the connection methods are considered internal?

**Reply by B4nan** [reply]
> This however doesn't seem to preserve the return type. Annotating the return type in a knex qb still gives me EntityData<Partial<any>>[] when using em.execute.

You can use the type argument to specify it explicitly, I guess I could try to make it infer the knex QB return type automatically, didn't really think about that before as it was mainly meant for executing raw queries which are just strings.

> Is there any way to close the connection after all unit tests without using em.getConnection().close() if the connection methods are considered internal?

It's on the `MikroORM` object, just `orm.close()`.

https://mikro-orm.io/docs/guide/project-setup#testing-the-endpoint
